### PR TITLE
feat(query): record copied files for insert with stage

### PR DESCRIPTION
**Pull request by everpcpc**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

after https://github.com/datafuselabs/databend/pull/10427

Closes #issue

### CODE REVIEW

File: src/query/service/src/interpreters/interpreter_insert.rs [hunk 1]
```diff
@@ -309,6 +334,20 @@ impl InsertInterpreter {
                     let append_entries = ctx.consume_precommit_blocks();
                     // We must put the commit operation to global runtime, which will avoid the "dispatch dropped without returning error" in tower
                     GlobalIORuntime::instance().block_on(async move {
+                        let mut copied_files = BTreeMap::new();
+                        for file in need_copy_files {
+                            // Short the etag to 7 bytes for less space in metasrv.
```
Comment (Xuanwo) [inline src/query/service/src/interpreters/interpreter_insert.rs:339]: This doesn't seem safe.

Comment (everpcpc) [reply src/query/service/src/interpreters/interpreter_insert.rs:339]: cc @youngsofun

Comment (BohuTANG) [reply src/query/service/src/interpreters/interpreter_insert.rs:339]: Can you convert the `insert ... attach file` to `copy`?

Comment (youngsofun) [reply src/query/service/src/interpreters/interpreter_insert.rs:339]: copy will record the file metadata in the meta server.
insert_with_attach  is an internal optimization for insert, 
the file is temporary.

**Comment by youngsofun** [comment]
@everpcpc when does a user need  `insert with stage `  instead of copy?

**Comment by everpcpc** [comment]
> @everpcpc when does a user need `insert with stage ` instead of copy?

Users do not use insert with stage directly, they use insert with batch instead, and we insert the batched values with stage in our drivers, without changing the user sql statement.

**Comment by flaneur2020** [comment]
> @everpcpc when does a user need `insert with stage ` instead of copy?

Insert with stage mainly targets the users on writing application with our SDK, likewise you can write the following code in snowflake sdk:

```
// Create a table containing an integer, float, boolean, and string column.
_, err = db.Exec("create or replace table my_table(c1 int, c2 float, c3 boolean, c4 string)")
...
// Define the arrays containing the data to insert.
intArray := []int{1, 2, 3}
fltArray := []float64{0.1, 2.34, 5.678}
boolArray := []bool{true, false, true}
strArray := []string{"test1", "test2", "test3"}
...
// Insert the data from the arrays and wrap in an Array() function into the table.
_, err = db.Exec("insert into my_table values (?, ?, ?, ?)", Array(&intArray), Array(&fltArray), Array(&boolArray), Array(&strArray))
```

all the user need is prepare the data batch in their favorite language.

on the other hand, the snowflake sdk will batch these data into a temp stage file in the background, the users do not need to care about the details on 1. presign a staged file; 2. upload a staged file; 3. copy into, just a insert with placeholders.

https://pkg.go.dev/github.com/snowflakedb/gosnowflake#hdr-Binding_Parameters_to_Array_Variables

**Comment by youngsofun** [comment]
so we do not need to `record copied files`?

https://github.com/datafuselabs/databend/pull/10661#discussion_r1142891637

**Comment by flaneur2020** [comment]
I guess there's still some benefit on allowing `insert with stage` to be idempotent.

like the following:

1. users uploaded the file to stage
2. insert with stage
3. <-- network lost between sdk and databend-query -->

if we got the network issue in step 3, the client cannot get whether the file have ingested successfully or not, a retry is not feasible without the idempotent semantic, as may causes duplications.

**Comment by youngsofun** [comment]
for now,  we do not support ErrorMode of `skip by file`
if tmp files are always purged,
we can check if these files exist (either all exists, or all purged).

**Comment by youngsofun** [comment]
if tmp files are always purged, client can remember the query_id, just check if it  has succeed is enough.
even when the conn is lost

**Comment by flaneur2020** [comment]
> if tmp files are always purged, client can remember the query_id, just check if it has succeed is enough. even when the conn is lost
> 
> if failed, client may choose to retry

this is a feasible approach, we can also get idempotent by remembering the recent committed query ids. I believe idempotent by query id is a more lightweight solution than remembering the file names.

how do we remember the recent committed query ids in databend right now?

**Comment by youngsofun** [comment]
history？

**Comment by flaneur2020** [comment]
can we guarantee the idempotent semantic by query id at the current version of databend? 🤔

**Comment by youngsofun** [comment]
> can we guarantee the idempotent semantic by query id at the current version of databend? 🤔

Thus, the probability to find a duplicate within 103 trillion version-4 UUIDs is one in a billion.


https://en.wikipedia.org/wiki/Universally_unique_identifier#Collisions

**Comment by flaneur2020** [comment]
> 

UUID collision is not a problem, but we have to remember the committed query id in some durable storage with some transaction to guarantee the idempotent semantic.

File: src/query/service/src/interpreters/interpreter_insert.rs (hunk 1 above)
Comment (flaneur2020) [reply src/query/service/src/interpreters/interpreter_insert.rs]: i guess the insert ... attach file can share the same logic with the `copy` statement ?

**Comment by youngsofun** [comment]
> > 
> 
> UUID collision is not a problem, but we have to remember the committed query id in some durable storage with some transaction to guarantee the idempotent semantic.

yes,  currently history is not persisted? it should be fixed since there may be other things that depend on it? 
cc @BohuTANG

File: src/query/service/src/interpreters/interpreter_insert.rs (hunk 1 above)
Comment (youngsofun) [reply src/query/service/src/interpreters/interpreter_insert.rs]: yes, but the detail depends on whether we  `need to record copied files`

**Comment by BohuTANG** [comment]
Hi, @everpcpc 
What's the progress of this PR, we do not seem to be working on it anymore, ready to close?

**Pull request closed without merging**